core/src/main/java/io/grpc/internal/AbstractClientStream.java

```diff
@@ -119,8 +124,14 @@ protected void inboundHeadersReceived(Metadata.Headers headers) {
     }
     if (headers.containsKey(ChannelImpl.MESSAGE_ENCODING_KEY)) {
       String messageEncoding = headers.get(ChannelImpl.MESSAGE_ENCODING_KEY);
-      setDecompressor(messageEncoding);
+      try {
+        setDecompressor(messageEncoding);
+      } catch (IllegalArgumentException e) {
+        inboundTransportError(Status.INVALID_ARGUMENT.withDescription(e.getMessage()));
```

Comment:
This is invalid input, since the remote host tried to use a message encoding that the local one does not support.

cancel() cannot be called, since the javadoc says it can only be called from the application, not the transport.  If there is some sort of inboundStreamError (just enough to kill the call) that would be the bets choice IMO.